### PR TITLE
Fix shortcuts on Dvorak–QWERTY Command layouts

### LINKED ISSUE #7511: ⌘ shortcuts broken on macOS "Dvorak – QWERTY ⌘" layout: letter matching uses unmodified Dvorak characters instead of the Command plane
### cmux version

0.64.17

### macOS version

macOS 26.2 (build 25C56)

### Mac chip

Apple Silicon (M1/M2/M3/M4)

### Installation method

Direct download (DMG)

### Can you reproduce this on cmux NIGHTLY?

I could not test NIGHTLY

### Bug description

I use the macOS built-in **"Dvorak – QWERTY ⌘"** input source (`KeyboardLayout Name = "DVORAK - QWERTY CMD"`). This layout types Dvorak normally, but it defines a **Command modifier plane**: while ⌘ is held, keys produce their **QWERTY** characters. The whole point is that ⌘-shortcut muscle memory stays at physical QWERTY positions in every app while typing stays Dvorak. Native AppKit apps (Finder, Safari, Terminal.app, and Ghostty upstream) honor this automatically because the OS-resolved event characters already go through the Command plane.

In cmux, ⌘+letter shortcuts don't follow this plane. Shortcuts behave as if bound to the **Dvorak** character positions (the layout's *unmodified* layer) instead of the QWERTY characters macOS reports while ⌘ is held. So pressing a shortcut at its physical QWERTY position — which is exactly what this layout is designed for — either does nothing or triggers a different action, and the action fires from the Dvorak character position instead.

Concrete example of the mechanism (Dvorak has "j" on the physical QWERTY-C key):

- I press ⌘ + physical C key. macOS resolves this to **⌘C** (Command plane = QWERTY), and that's what I intend: copy.
- cmux's matcher translates the keyCode through the unmodified layer and sees **"j"**, so the ⌘C binding doesn't match.

**Root cause hypothesis:** #762 fixed plain-Dvorak users by requiring a character match for ⌘+letter shortcuts, using the unmodified-layer translation added in #680 (`KeyboardLayout.character(forKeyCode:)`). For plain Dvorak the unmodified layer is the right answer, but for "Dvorak – QWERTY ⌘" the unmodified layer (Dvorak) and the Command plane (QWERTY) disagree *precisely when ⌘ is held* — so that fix inverts the behavior for this layout. Before #762, physical-keyCode matching happened to work correctly for this layout.

Note the tension: plain-Dvorak users need character matching, Dvorak–QWERTY⌘ users need Command-plane matching. Both are satisfied by matching against the **OS-resolved characters of the event** (AppKit already applies the Command plane), or by translating keycodes with the ⌘ modifier state applied (`UCKeyTranslate` with `cmdKey`) rather than with no modifiers. The physical-keyCode-primary refactor proposed in #2215 / #2216 would also fix this layout as a side effect.

### Expected behavior

With "Dvorak – QWERTY ⌘" active, all ⌘-modified shortcuts match at their physical QWERTY positions — identical to native macOS apps and upstream Ghostty. Unmodified typing stays Dvorak.

### Steps to reproduce

1. System Settings → Keyboard → Input Sources → add **"Dvorak – QWERTY ⌘"** (under English) and switch to it.
2. Open cmux.
3. Press ⌘+letter shortcuts at their physical QWERTY positions (e.g. ⌘T for new surface, ⌘C to copy).
4. Observe they don't fire (or a different action fires). Pressing the key at the Dvorak character position triggers the binding instead.

### Shell and environment

zsh. Input source verified via `defaults read com.apple.HIToolbox AppleEnabledInputSources` → `"KeyboardLayout Name" = "DVORAK - QWERTY CMD"`.

### Relevant logs or crash reports

_No response_

### Screenshots or screen recordings

_No response_

### Additional context

Related issues/PRs: #762 (plain-Dvorak character matching — the change that inverts behavior for this layout), #680 (unmodified-layer keycode translation), #2215 / #2216 (physical keyCode as primary match path, like upstream Ghostty), #3362 (alternative layouts not considered for hotkeys), #5810 (zoom shortcuts on non-US layouts).

"Dvorak – QWERTY ⌘" ships with macOS and is the layout Apple provides specifically so Dvorak typists keep QWERTY shortcuts — it's a common setup among Dvorak users, and the same bug class previously affected Chromium/Electron for years, so a regression test for this layout may be worthwhile alongside the plain-Dvorak tests added in #762.

### PR DESCRIPTION
**Pull request by austinywang**
## Summary

- Resolve Command shortcuts from AppKit modifier-aware event characters so Dvorak–QWERTY ⌘ uses its QWERTY Command plane.
- Reuse one event-scoped `ShortcutEventCharacterResolver` across stored bindings, command palette routing, global search, sidebar modes, and screenshot cancellation.
- Preserve the intended base key for Command+Option/Control shortcuts by translating the exact Command/Shift plane before using the ASCII-capable fallback.

Fixes #7511

## Testing

- Red proof: [hosted run 30886812511](https://github.com/manaflow-ai/cmux/actions/runs/30886812511) checked out test-only commit `b0e8893d5b4d69a9696e9b3282695fb9abf859f0`; the new Dvorak–QWERTY Command regression failed with `"y" != "t"` and rejected Cmd+T.
- Green proof: [hosted run 30887675353](https://github.com/manaflow-ai/cmux/actions/runs/30887675353) passed the focused regression after the implementation landed.
- Final-head fleet proof: run `sym7511-tests9-e484fe733958` tested commit `a4afd132bcdc351ce613c47f38b61b05259e3404` with the `cmux-unit` scheme and five focused `-only-testing` filters; all 14 tests in 5 suites passed. A broader run exposed the existing global-search chord failure, and exact pre-PR baseline run `sym7511-basechord-3540f994f8e7` failed at the same assertion on the same fleet host.
- Project guards: `./scripts/check-pbxproj.sh`, `./scripts/lint-pbxproj-test-wiring.sh`, `python3 scripts/check-package-resolved-policy.py`, and `git diff --check f4cd2774de..HEAD`.
- Tagged dev build: `./scripts/reload-cloud.sh --builder fleet --host cmux9s-mac-mini --tag sym7511 --launch` built exact final commit `a4afd132bcdc351ce613c47f38b61b05259e3404` as fleet run `sym7511-496d68181db1`; no local fallback was used.
- Dogfood on `/tmp/cmux-debug-sym7511.sock`: with `com.apple.keylayout.DVORAK-QWERTYCMD` active, the app had 5 surfaces; simulated Cmd+physical-QWERTY-Y left it at 5, while Cmd+physical-QWERTY-T created and selected `surface:6` (6 total). Restored the U.S. layout, disabled the temporary Dvorak input source, quit with `pkill -f "DerivedData/cmux-sym7511"`, and confirmed the tagged process and socket were gone.
- Clean-head review: the canonical branch autoreview exited 0 with no findings at 0.93 confidence; the merge-conflict gate and cmux policy checks were clean.
- Localization audit: no user-facing strings, shortcut definitions, Settings schema, docs, localization resources, or message catalogs changed.

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/9537"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1788418111&installation_model_id=21920&pr_number=9537&repository=manaflow-ai%2Fcmux&return_to=https%3A%2F%2Fgithub.com%2Fmanaflow-ai%2Fcmux%2Fpull%2F9537&signature=14f648ec1754ed7b456ae91e80c7aa5a7c8cb71e939d1f77149f9db22aba85e2"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Fixes Command shortcuts on macOS Dvorak–QWERTY Command layouts by honoring AppKit’s Command-plane characters across the app. Cmd+T and similar now fire on the QWERTY Command layer, and Command+Option/Control keeps the intended base key. Fixes #7511.

- **Bug Fixes**
  - Use a per-event `ShortcutEventCharacterResolver` to trust Command-plane `characters` and cache an ASCII fallback; shared via AppDelegate’s focus-context and applied to shortcut matching, command palette (incl. field editor), right sidebar modes, global search, and screenshot overlay cancel (Cmd+.).
  - For Command+Option/Control, resolve the exact Command/Shift plane first; preserve non-ASCII/dead-key `charactersIgnoringModifiers`; skip translation on modifier mismatch or recorded key codes; ensure global search digits (Cmd+1–9) and system/editing commands read command-plane `characters`. Add tests for Dvorak–QWERTY ⌘, Command+Option base key, Unicode vs ASCII fallback, single resolver reuse across candidates, and field editor behavior.

- **Refactors**
  - Introduced `ShortcutEventCharacterResolver`; updated `KeyboardLayout.normalizedCharacters`, `KeyboardShortcutSettings` matching, and the focus-context cache to share one resolver per event and avoid redundant translations.

<sup>Written for commit a4afd132bcdc351ce613c47f38b61b05259e3404. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/9537?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved keyboard shortcut recognition across different keyboard layouts, including Dvorak and QWERTY.
  - Corrected matching for Command-Option shortcuts while preserving the intended physical key.
  - Improved handling of non-printable and layout-translated shortcut characters.

- **Tests**
  - Added regression coverage for keyboard-layout-specific shortcut matching.

### CODE REVIEW

File: cmuxTests/KeyboardShortcutContextSwiftTests.swift [hunk 1]
```diff
@@ -273,6 +274,63 @@ extension CMUXCLIErrorOutputRegressionTests {
...
+            isARepeat: false,
+            keyCode: 17
+        ))
+
+        #expect(KeyboardLayout.normalizedCharacters(for: event) == "t")
+        #expect(shortcut.matches(
+            event: event,
+            layoutCharacterProvider: { _, _ in "t" }
+        ))
+    }
```
[the remark this answers is not in the document]
Comment (austinywang) [reply cmuxTests/KeyboardShortcutContextSwiftTests.swift:323]: Addressed in f2f3e1dbd4: the Command-only provider now returns a differing character and records that it was not called; the Command-Option provider records and asserts the exact `.command` + `.option` flags.

**Comment by austinywang** [comment]
Follow-up on the stale CodeRabbit **Cmux No Ambient Global State** banner: current HEAD `a4afd132` no longer contains the named `KeyboardLayout.shortcutCharacters` method. The event-specific behavior now lives in the constructable, injected `ShortcutEventCharacterResolver`; the remaining `KeyboardLayout.normalizedCharacters` entry point delegates to that resolver and adds no ambient state. The canonical cmux policy check is clean on this exact HEAD.
